compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -1351,13 +1352,34 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                         );
                     } else {
                         let is_mut = mut_ref_self_ty_satisfies_pred || ref_inner_ty_mut;
+
+                        // see #104961
+                        let Some(body_id) = self.tcx.hir().maybe_body_owned_by(obligation.cause.body_id) else { return false; };
+                        let body = self.tcx.hir().body(body_id);
+                        let mut expr_finder = FindExprBySpan::new(span);
+                        expr_finder.visit_expr(body.value);
+                        let Some(expr) = expr_finder.result else { return false; };
+                        let needs_parens = match expr.kind {
+                            // parenthesize if needed (Issue #46756)
+                            hir::ExprKind::Cast(_, _) | hir::ExprKind::Binary(_, _, _) => true,
+                            // parenthesize borrows of range literals (Issue #54505)
+                            _ if is_range_literal(expr) => true,
+                            _ => false,
+                        };
+
+                        let span = if needs_parens { span } else { span.shrink_to_lo() };
+                        let sugg_prefix = format!("&{}", if is_mut { "mut " } else { "" });
                         err.span_suggestion_verbose(
-                            span.shrink_to_lo(),
+                            span,
                             &format!(
                                 "consider{} borrowing here",
                                 if is_mut { " mutably" } else { "" }
                             ),
-                            format!("&{}", if is_mut { "mut " } else { "" }),
+                            if !needs_parens {
+                                sugg_prefix
+                            } else {
+                                format!("{}({})", sugg_prefix, snippet)
```

Comment:
Suggestions that copy user code are typically hard to understand: everything is underlined, and the difference is not underlined.
Could you prefer a multipart suggestion with `&(` at the beginning and `)` at the end?

Reply:
make sense, fixed.